src/nnet3/nnet-simple-component.cc

```diff
@@ -136,34 +136,27 @@ void DropoutComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
 
   BaseFloat dropout = dropout_proportion_;
   KALDI_ASSERT(dropout >= 0.0 && dropout <= 1.0);
-  if(!dropout_per_frame_)
-  {
+  if (!dropout_per_frame_) {
     // This const_cast is only safe assuming you don't attempt
     // to use multi-threaded code with the GPU.
     const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(out);
 
-    out->Add(-dropout); // now, a proportion "dropout" will be <0.0
-    out->ApplyHeaviside(); // apply the function (x>0?1:0).  Now, a proportion "dropout" will
-                          // be zero and (1 - dropout) will be 1.0.
+    out->Add(-dropout);  // now, a proportion "dropout" will be <0.0
+    out->ApplyHeaviside();  // apply the function (x>0?1:0).  Now, a proportion
+                           // "dropout" will be zero and (1 - dropout) will be 1.0.
     out->MulElements(in);
   } else {
     // randomize the dropout matrix by row,
     // i.e. [[1,1,1,1],[0,0,0,0],[0,0,0,0],[1,1,1,1],[0,0,0,0]]
+    CuMatrix<BaseFloat> tmp(1, out->NumRows(), kUndefined);
     // This const_cast is only safe assuming you don't attempt
     // to use multi-threaded code with the GPU.
-    CuSubMatrix<BaseFloat> out_col_submatrix = out->ColRange(0, 1);
-    const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(&out_col_submatrix); 
-    // now, a proportion "dropout" will be <0.0
-    out_col_submatrix.Add(-dropout);
-    // apply the function (x>0?1:0).  Now, a proportion "dropout" will
-    // be zero and (1 - dropout) will be 1.0.
-    out_col_submatrix.ApplyHeaviside();
-    CuVector<BaseFloat> *random_drop_vector = new CuVector<BaseFloat>(in.NumRows(), kSetZero);
-    random_drop_vector->CopyColFromMat(out_col_submatrix, 0);
+    const_cast<CuRand<BaseFloat>&>(random_generator_).RandUniform(&tmp);
+    tmp.Add(-dropout);
+    tmp.ApplyHeaviside();
     out->SetZero();
-    out->AddVecToCols(1.0 , *random_drop_vector, 1.0);
+    out->AddVecToCols(1.0, tmp.Row(0), 0.0);
     out->MulElements(in);
```

Comment:
Change from my previous method (random a whole matrix and select one col) into this one, and a test on AMI SDM is done, they are equivalent on WER:
previous way:
![image](https://cloud.githubusercontent.com/assets/7532101/22152388/aedbe916-df5d-11e6-9743-e25c23951a60.png)
new way:
![image](https://cloud.githubusercontent.com/assets/7532101/22152392/b84f4a88-df5d-11e6-8b9f-7c5286910a6b.png)